File: 0853-car-fleet/0853-car-fleet.py
"""
Time = O(n)
Space = O(n)
"""
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        pos_speed = []
        for i in range(len(position)):
            pos_speed.append((position[i], speed[i]))
        
        # Sort the list
        pos_speed.sort()
        
        stack = []
        
        # Traverse the pos_speed list reversely
        for pos, speed in pos_speed[::-1]:
            # Calculate the time needed for current car to reach target
            time = (target - pos) / speed
            # Found car that can reach the car in its front
            if not stack or time > stack[-1]:
                stack.append(time)
        
        return len(stack)
        

File: 0853-car-fleet/0853-car-fleet.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        vector<pair<int, double>> posSpeedPairs;
        stack<double> stk;
        
        // make pair
        for (int i = 0; i < speed.size(); i++) {
            double time = (double) (target - position[i]) / speed[i];
            posSpeedPairs.push_back(make_pair(position[i], time));
        }
        
        // sort by position
        sort(posSpeedPairs.begin(), posSpeedPairs.end());

        for (int i = posSpeedPairs.size() - 1; i >= 0; i--) {
            double time = posSpeedPairs[i].second;
            // cannot chase the front car
            if(stk.empty() || time > stk.top()) {
                stk.push(posSpeedPairs[i].second);
            }
        }
        // number of car fleets
        return stk.size();
    }
};